File: scripts/update_compat_report_issue.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.14"
# ///

"""Create or update the managed Shields.io compatibility report issue."""

from __future__ import annotations

import argparse
import json
import os
import subprocess
import sys
from pathlib import Path


def run_gh(args: list[str]) -> subprocess.CompletedProcess[str]:
    return subprocess.run(
        ["gh", *args],
        text=True,
        capture_output=True,
        check=False,
    )


def warn(message: str) -> None:
    print(f"::warning::{message}", file=sys.stderr)


def ensure_label(repo: str, label: str) -> bool:
    result = run_gh(
        [
            "label",
            "create",
            label,
            "--repo",
            repo,
            "--description",
            "Managed Shields.io compatibility report issue",
            "--color",
            "5319e7",
        ]
    )
    if result.returncode == 0:
        return True

    message = result.stderr.strip() or result.stdout.strip()
    if "already exists" in message.lower():
        return True

    warn(f"could not create compatibility report label: {message}")
    return False


def find_issue(repo: str, label: str) -> int | None:
    result = run_gh(
        [
            "issue",
            "list",
            "--repo",
            repo,
            "--state",
            "open",
            "--author",
            "github-actions[bot]",
            "--label",
            label,
            "--json",
            "number,updatedAt",
        ]
    )
    if result.returncode != 0:
        message = result.stderr.strip() or result.stdout.strip()
        warn(f"could not search compatibility report issues: {message}")
        return None

    issues = json.loads(result.stdout)
    if not issues:
        return None

    issues.sort(key=lambda issue: issue["updatedAt"], reverse=True)
    return int(issues[0]["number"])


def issue_url(repo: str, number: int) -> str:
    result = run_gh(
        [
            "issue",
            "view",
            str(number),
            "--repo",
            repo,
            "--json",
            "url",
        ]
    )
    if result.returncode != 0:
        return f"https://github.com/{repo}/issues/{number}"
    return str(json.loads(result.stdout)["url"])


def create_issue(repo: str, label: str, title: str, body_file: Path) -> str | None:
    result = run_gh(
        [
            "issue",
            "create",
            "--repo",
            repo,
            "--title",
            title,
            "--label",
            label,
            "--body-file",
            str(body_file),
        ]
    )
    if result.returncode != 0:
        message = result.stderr.strip() or result.stdout.strip()
        warn(f"could not create compatibility report issue: {message}")
        return None
    return result.stdout.strip()


def update_issue(repo: str, number: int, body_file: Path) -> str | None:
    result = run_gh(
        [
            "issue",
            "edit",
            str(number),
            "--repo",
            repo,
            "--body-file",
            str(body_file),
        ]
    )
    if result.returncode != 0:
        message = result.stderr.strip() or result.stdout.strip()
        warn(f"could not update compatibility report issue #{number}: {message}")
        return None
    return issue_url(repo, number)


def write_output(name: str, value: str) -> None:
    github_output = os.environ.get("GITHUB_OUTPUT")
    if not github_output:
        return
    with Path(github_output).open("a", encoding="utf-8") as output:
        print(f"{name}={value}", file=output)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--repo", required=True, help="GitHub repository, owner/name")
    parser.add_argument("--label", required=True, help="Dedicated tracking label")
    parser.add_argument("--title", required=True, help="Issue title to create")
    parser.add_argument(
        "--body-file",
        required=True,
        type=Path,
        help="Markdown report body",
    )
    args = parser.parse_args()

    if not args.body_file.is_file():
        raise SystemExit(f"report body does not exist: {args.body_file}")

    if not ensure_label(args.repo, args.label):
        return 0

    number = find_issue(args.repo, args.label)
    if number is None:
        url = create_issue(args.repo, args.label, args.title, args.body_file)
    else:
        url = update_issue(args.repo, number, args.body_file)

    if url:
        print(f"compatibility report issue: {url}")
        write_output("issue_url", url)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/update_compat_report_issue.js
#!/usr/bin/env bun

import { appendFileSync, existsSync, readFileSync } from "node:fs";
import { spawnSync } from "node:child_process";

function usage() {
  return "usage: update_compat_report_issue.js --repo <owner/repo> --label <label> --title <title> --body-file <path>";
}

function parseArgs(argv) {
  if (argv.includes("--help") || argv.includes("-h")) {
    console.log(usage());
    process.exit(0);
  }

  const args = {};
  for (let index = 0; index < argv.length; index += 2) {
    const key = argv[index];
    const value = argv[index + 1];
    if (!key?.startsWith("--") || value === undefined) {
      throw new Error(usage());
    }
    args[key.slice(2)] = value;
  }

  for (const required of ["repo", "label", "title", "body-file"]) {
    if (!args[required]) {
      throw new Error(usage());
    }
  }
  return args;
}

function runGh(args) {
  return spawnSync("gh", args, {
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
  });
}

function commandMessage(result) {
  return result.stderr.trim() || result.stdout.trim();
}

function warn(message) {
  console.error(`::warning::${message}`);
}

function ensureLabel(repo, label) {
  const result = runGh([
    "label",
    "create",
    label,
    "--repo",
    repo,
    "--description",
    "Managed Shields.io compatibility report issue",
    "--color",
    "5319e7",
  ]);
  if (result.status === 0) {
    return true;
  }

  const message = commandMessage(result);
  if (message.toLowerCase().includes("already exists")) {
    return true;
  }

  warn(`could not create compatibility report label: ${message}`);
  return false;
}

function findIssue(repo, label) {
  const result = runGh([
    "issue",
    "list",
    "--repo",
    repo,
    "--state",
    "open",
    "--author",
    "github-actions[bot]",
    "--label",
    label,
    "--json",
    "number,updatedAt",
  ]);
  if (result.status !== 0) {
    warn(`could not search compatibility report issues: ${commandMessage(result)}`);
    return undefined;
  }

  const issues = JSON.parse(result.stdout);
  if (issues.length === 0) {
    return undefined;
  }

  issues.sort((left, right) => right.updatedAt.localeCompare(left.updatedAt));
  return issues[0].number;
}

function issueUrl(repo, number) {
  const result = runGh([
    "issue",
    "view",
    String(number),
    "--repo",
    repo,
    "--json",
    "url",
  ]);
  if (result.status !== 0) {
    return `https://github.com/${repo}/issues/${number}`;
  }
  return JSON.parse(result.stdout).url;
}

function createIssue(repo, label, title, bodyFile) {
  const result = runGh([
    "issue",
    "create",
    "--repo",
    repo,
    "--title",
    title,
    "--label",
    label,
    "--body-file",
    bodyFile,
  ]);
  if (result.status !== 0) {
    warn(`could not create compatibility report issue: ${commandMessage(result)}`);
    return undefined;
  }
  return result.stdout.trim();
}

function updateIssue(repo, number, bodyFile) {
  const result = runGh([
    "issue",
    "edit",
    String(number),
    "--repo",
    repo,
    "--body-file",
    bodyFile,
  ]);
  if (result.status !== 0) {
    warn(`could not update compatibility report issue #${number}: ${commandMessage(result)}`);
    return undefined;
  }
  return issueUrl(repo, number);
}

function writeOutput(name, value) {
  const githubOutput = process.env.GITHUB_OUTPUT;
  if (!githubOutput) {
    return;
  }
  appendFileSync(githubOutput, `${name}=${value}\n`, "utf8");
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const bodyFile = args["body-file"];
  if (!existsSync(bodyFile)) {
    throw new Error(`report body does not exist: ${bodyFile}`);
  }

  // Touch the body file early so missing permissions or bad paths fail before gh calls.
  readFileSync(bodyFile, "utf8");

  if (!ensureLabel(args.repo, args.label)) {
    return;
  }

  const number = findIssue(args.repo, args.label);
  const url =
    number === undefined
      ? createIssue(args.repo, args.label, args.title, bodyFile)
      : updateIssue(args.repo, number, bodyFile);

  if (url) {
    console.log(`compatibility report issue: ${url}`);
    writeOutput("issue_url", url);
  }
}

try {
  main();
} catch (error) {
  console.error(error instanceof Error ? error.message : String(error));
  process.exit(1);
}
